### PR TITLE
MNIST params

### PR DESCRIPTION
**Pull request by breznak**
- MNIST experiments with manual configuration
  - found better performance
  - and much smaller, runnable model
- SP cleanup  
  - (OT: I'd like to head to removing boosting) 

TODO:
 - [ ] copy the best params over to python

In OTHER PR:
- make py mnist runnable out of the box (use c++'s MNIST data)
- mnist now runs fast enough -> make it a test same as hotgym is

### CODE REVIEW

File: src/examples/mnist/MNIST_SP.cpp [hunk 1]
```diff
@@ -31,16 +31,32 @@
...
 
 class MNIST {
+/**
+ * RESULTS:
+ *
+ * Order :	score				: column dimensions 	: #passes on train	: time to run (s)	: git commit	: comment
+ * -------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+ * 1/	Score: 97.11% (289 / 10000 wrong)	: 28x28x16		: 4			: 557			: 1f0187fc6 	: epochs help, at cost of time 
+ *
+ * 2/	Score: 96.56% (344 / 10000 wrong)	: 28x28x16		: 1			: 142			: 3ccadc6d6
```
Comment (breznak) [inline src/examples/mnist/MNIST_SP.cpp]: competitive results, considering it's single pass, and many "ML" features missing (dropout, data-modification, cross-validation, param opt, many training passes...)

File: src/examples/mnist/MNIST_SP.cpp [hunk 2]
```diff
@@ -51,31 +67,31 @@ class MNIST {
...
+    /* potentialPct */                0.1f, //we have only 10 classes, and << #columns. So we want to force each col to specialize. Cca 0.3 w "7" above, or very small (0.1) for "no topology". Cannot be too small due to internal checks. Speed++
+    /* globalInhibition */            true, //Speed+++++++; SDR quality-- (global does have active nearby cols, which we want to avoid (local)); Results+-0
+    /* localAreaDensity */            0.1f,  // % active bits
     /* numActiveColumnsPerInhArea */  -1,
     /* stimulusThreshold */           6u,
-    /* synPermInactiveDec */          0.005f,
-    /* synPermActiveInc */            0.014f,
-    /* synPermConnected */            0.1f,
-    /* minPctOverlapDutyCycles */     0.001f,
+    /* synPermInactiveDec */          0.002f, //FIXME inactive decay permanence plays NO role, investigate! (slightly better w/o it)
```
Comment (breznak) [inline src/examples/mnist/MNIST_SP.cpp]: I never get to decay play a role!?

File: src/htm/algorithms/SpatialPooler.cpp [hunk 3]
```diff
@@ -755,6 +760,16 @@ void SpatialPooler::updateBoostFactors_() {
 }
 
 
+void applyBoosting_(const UInt i,
+		    const Real targetDensity, 
+		    const vector<Real>& actualDensity,
+		    const Real boost,
+	            vector<Real>& output) {
+  if(boost < htm::Epsilon) return; //skip for disabled boosting
+  output[i] = exp((targetDensity - actualDensity[i]) * boost); //TODO doc this code
```
Comment (breznak) [inline src/htm/algorithms/SpatialPooler.cpp]: where is a doc for this exp based boosting, please?

File: src/htm/algorithms/SpatialPooler.cpp [hunk 4]
```diff
@@ -500,6 +500,11 @@ void SpatialPooler::compute(const SDR &input, const bool learn, SDR &active) {
 
 void SpatialPooler::boostOverlaps_(const vector<SynapseIdx> &overlaps, //TODO use Eigen sparse vector here
                                    vector<Real> &boosted) const {
+  if(boostStrength_ < htm::Epsilon) {
+    const auto begin = static_cast<const SynapseIdx*>(overlaps.data());
+    boosted.assign(begin, begin + overlaps.size());
+    return; //boost ~ 0.0, we can skip these computations.
```
Comment (breznak) [inline src/htm/algorithms/SpatialPooler.cpp]: this probably changed the results for the test

File: src/examples/mnist/MNIST_SP.cpp (hunk 1 above)
Comment (ctrl-z-9000-times) [reply src/examples/mnist/MNIST_SP.cpp]: I literally can not read this table because the lines are too long.

Comment (breznak) [reply src/examples/mnist/MNIST_SP.cpp]: better now?

**Comment by breznak** [comment]
Can we merge this? I'll use it for testing the other PRs

File: src/htm/algorithms/SpatialPooler.cpp [hunk 5]
```diff
@@ -172,7 +172,7 @@ void SpatialPooler::setDutyCyclePeriod(UInt dutyCyclePeriod) {
 Real SpatialPooler::getBoostStrength() const { return boostStrength_; }
 
 void SpatialPooler::setBoostStrength(Real boostStrength) {
-  NTA_CHECK(boostStrength >= 0.0f);
+  NTA_CHECK(boostStrength == 0.0f or boostStrength >= 1.0) << "Boost strength must be >= 1.0, or exactly 0.0 (=disabled).";
```
Comment (ctrl-z-9000-times) [inline src/htm/algorithms/SpatialPooler.cpp]: Why is this a restriction?  Does it break with `0 < boostStrength < 1` ?

File: src/htm/algorithms/SpatialPooler.cpp (hunk 4 above)
Comment (ctrl-z-9000-times) [reply src/htm/algorithms/SpatialPooler.cpp]: Can you be a little more certain of this?  Why exactly did the unit tests change?

Comment (ctrl-z-9000-times) [reply src/htm/algorithms/SpatialPooler.cpp]: Nevermind, I see how your unit test works.

File: src/htm/algorithms/SpatialPooler.cpp (hunk 5 above)
Comment (breznak) [reply src/htm/algorithms/SpatialPooler.cpp]: I can remove this, my idea was that if we boost, we want to do so in a positive way, not "silencing" the column. 
Actually, I'd like to move to "your" log-boost function that is parameterless.

Comment (breznak) [reply src/htm/algorithms/SpatialPooler.cpp]: reverted these limits. It could make sense to use 0 < strength < 1, to "soften" the boosting.

File: src/htm/algorithms/SpatialPooler.cpp (hunk 4 above)
Comment (ctrl-z-9000-times) [reply src/htm/algorithms/SpatialPooler.cpp]: nitpick: `boosted.assign(overlaps.begin(), overlaps.end())`

File: src/htm/algorithms/SpatialPooler.hpp [hunk 6]
```diff
@@ -185,8 +185,9 @@ class SpatialPooler : public Serializable
         boost. Shorter values make it potentially more unstable and
         likely to oscillate.
 
-  @param boostStrength A number greater or equal than 0, used to
-        control boosting strength. No boosting is applied if it is set to 0.
+  @param boostStrength A number greater or equal than 1.0 or exactly 0.0, used to
```
Comment (ctrl-z-9000-times) [inline src/htm/algorithms/SpatialPooler.hpp]: Please revert this change to the docs, since you reverted the corresponding code change.

Comment (breznak) [reply src/htm/algorithms/SpatialPooler.hpp]: good catch, reverted. Thanks!

**Comment by ctrl-z-9000-times** [comment]
> copy the best params over to python

Please do not do this.  The python version already has good parameters.  IIRC it scores above 95% after a single iteration through the training data.

**Comment by breznak** [comment]
> Please do not do this. The python version already has good parameters. IIRC it scores above 95% after a single iteration through the training data.

ok, I won't now. Uups, should have looked there for bootstrap params! 
But, one thing I notice there, it uses quite large (~35k) # of columns (and so probably runs slow, on my ntb MNIST before this PR is cca 10mins, now 140s!). 
I noticed this and it's one thing we need to consider and should add into the param optimization: with large enough number of columns we can solve any problem (Kolmogorov), but we'll end up with a large network of dummy neurons. 
I noticed that with some params, where such network would run faster -> allowing me to increase the number of columns for the same training time. I try to avoid this to some reasonable extent.
